DbgShell/ConsoleControl.AnsiColorWriter.cs

```diff
@@ -140,253 +152,425 @@ public AnsiColorWriter()
+                index++;
+                if( 0 == code )
+                {
+                    //using default foreground/background codes instead of full reset so that the colorcodestack can maintain either one as default independently
+                    ForegroundCode = new VtColorCode( 39 );
+                    BackgroundCode = new VtColorCode( 49 );
+                }
+                else if( ((code <= 37) && (code >= 30)) || ((code <= 97) && (code >= 90)) || code == 39 )
+                {
+                    ForegroundCode = new VtColorCode( code );
+                }
+                else if( ((code <= 47) && (code >= 40)) || ((code <= 107) && (code >= 100)) || code == 49 )
+                {
+                    BackgroundCode = new VtColorCode( code );
+                }
+                else if( code == 38 )
+                {
+                    ForegroundCode = _ProcessTrueColorSgrCode( code, codes, ref index );
+                }
+                else if( code == 48 )
+                {
+                    BackgroundCode = _ProcessTrueColorSgrCode( code, codes, ref index );
+                }
+                else if( 56 == code ) // NON-STANDARD (I made this one up)
+                {
+                    Util.Fail( "The 56/57 SGR codes (non-standard push/pop) are deprecated. Use XTPUSHSGR/XTPOPSGR instead." );
+                    m_colorCodeStack.Push( new ColorCodePair( ForegroundCode, BackgroundCode ) );
+                }
+                else if( 57 == code ) // NON-STANDARD (I made this one up)
+                {
+                    Util.Fail( "The 56/57 SGR codes (non-standard push/pop) are deprecated. Use XTPUSHSGR/XTPOPSGR instead." );
+                    var pair = m_colorCodeStack.Pop();
+                    ForegroundCode = pair.Foreground;
+                    BackgroundCode = pair.Background;
+                }
+                else
+                {
+                    throw new NotSupportedException( String.Format( "SGR code '{0}' not supported.", code ) );
+                }
+            } // end _ProcessSgrCode()
+
+            private static VtColorCode _ProcessTrueColorSgrCode( int command,  List< int > codes, ref int index )
             {
-                throw new NotSupportedException( String.Format( "SGR code '{0}' not supported.", code ) );
+                if( codes.Count - index < 4 ) { throw new ArgumentException( "Extended SGR sequence does not have enough arguments for RGB colors" ); }
+
+                if( codes[index] != 2 ) { throw new NotSupportedException( $"Extended SGR mode '{codes[ index + 1 ]}' not supported." ); }
+                index++;
+                return new VtColorCode( (byte)command, (byte) codes[ index++ ], (byte) codes[ index++ ], (byte) codes[ index++ ] );
             }
-        } // end _ProcessSgrCode()
-    } // end class AnsiColorWriter
-} // class ConsoleControl
+
+            //[zhent] Because the cost of .ToString() is just too much to bear.
```

Comment:
>Because the cost of .ToString() is just too much to bear. [](start = 22, length = 57)

![fast](https://media.giphy.com/media/ulZwxYQVqUWhW/giphy.gif)